### DISCUSSION TITLE
Added Mako to TechEmPower Framework Benchmarks

### DISCUSSION THREAD

**Discussion by n0nag0n** [mako-framework/framework#304] [Show and tell]
Hey there!

Came across your framework when browser through php frameworks on Github. I wanted to mess with it, but didn't have a project, so I just threw it in the benchmark as "my project". https://github.com/TechEmpower/FrameworkBenchmarks/pull/7827

Hopefully it gets pulled in soon. From performance testing on my machine, it seems it is about as fast as Fat-Free, which is quite impressive! I enjoyed working with it and will probably be using it on a next project of mine. I will say, I prefer Latte as the templating engine instead of the Twig like one you had in there. I saw the docs for how to extend the templating engine (though I am a little confused on where you put that extend() method in which file...) 

Anywho, great framework! Thank you for all your hard work! @freost

**Comment by n0nag0n** [comment]
Btw, one thing I really would love about the documentation is a search feature. I'm pretty sure Algolia would be free for open source projects, and if I remember right, I think it's as simple as point a URL at it, and it will index it for you. https://www.algolia.com/for-open-source/

**Reply by freost** [reply]
I finally got around to adding Algolia search to the docs. I still need to tweak the crawler docs though.

**Reply by n0nag0n** [reply]
Whoa!!! That's awesome! Looks great!

**Comment by freost** [comment]
Hi! 

Thanks for the feedback. A combination of performance and ease of use has been the main goal so I'm glad you like it. I agree that the documentation needs a search feature so I'll look into setting up Algolia sometime soon.

A good place to set up a custom template renderer would be in the `bootstrap.php` file (I'll update the docs):

```php
<?php

/** @var \mako\syringe\Container $container */

use mako\view\ViewFactory;

$container->get(ViewFactory::class)->extend('.latte', function()
{

});
```

**Reply by n0nag0n** [reply]
Ahh cool, thanks!

**Comment by n0nag0n** [comment]
@freost Looks like your framework performs about as well as Fat-Free which I think is a huge win! Great framework with good amount of features and it performs very very well. Congrats! https://www.techempower.com/benchmarks/#section=test&runid=dc15ab19-0139-4489-b767-36eab8443eb1&test=plaintext

A follow up question would be if you know if your framework is compatible with other async backend solutions like swoole, roadrunner, workerman, etc?

And finally, do you guys have a discord or slack channel? It'd be way easier to hit you up on that instead of this haha. I have a bunch of guys at my work that are becoming very interested in Mako

**Reply by freost** [reply]
That's pretty cool. Thanks for adding Mako to the benchmark test suite!

I don't think think that it'll be compatible right out of the box but I have been thinking of trying to get it running on one of them. I just haven't had time to look at it yet.

There's no Slack or Discord at the moment but I might set one up in the future. For now I'll do my best to answer any questions here on the discussion board.

**Reply by n0nag0n** [reply]
@freost If you don't have something, join in using this link [https://matrix.to/#/#mako-framework:matrix.org](https://matrix.to/#/#mako-framework:matrix.org). I use app.element.io but there's a lot of other apps you can use to access the Matrix decentralized framework. We chose this because there's no cost, no limits of history (unlike slack), and people without mobile numbers can login just fine (unlike discord). I've already created it. So if you hop in and say hi, I should be there and I can make you admin. If you like it, then you can toss the invite link around the website so others can join in if they want.

**Reply by n0nag0n** [reply]
Here's a bunch of clients that can access it. https://matrix.org/clients/

**Comment by n0nag0n** [comment]
They finally made a public new Round and Mako made it on the list. Nice job for not having swoole/workerman! @freost https://www.techempower.com/benchmarks/#hw=ph&test=plaintext&section=data-r22&l=zik073-cn3

**Reply by freost** [reply]
Thanks! I'm pretty happy with the results and I think Mako 10 is slightly faster than 9 thanks to some changes in the middleware/constraint/dispatcher code.

I have also been toying with the idea of making some minor changes so that Mako can run on [FrankenPHP](https://frankenphp.dev/). It should provide a decent boost in performance since application core services, routes and database connections won't have to be initialized for every request.

**Reply by n0nag0n** [reply]
Hey that's cool! I'm sure when you make those changes, it'll allow it to work with Swoole, Workerman, and others. FrankenPHP is cool, but based on Go and while it's fast, it's just not the fastest.

Great work though!

**Comment by freost** [comment]
The initial FrankenPHP tests seem promising. This is just a simple "Hello, world" test after spending a few minutes on some minor changes to the framework to make it compatible with the worker mode and it runs about 4 times faster on FrankenPHP.

Mako 10 on Apache + PHP 8.2 (using opcache production settings):

```
Concurrency Level:      100
Time taken for tests:   0.471 seconds
Complete requests:      1000
Failed requests:        0
Total transferred:      206000 bytes
HTML transferred:       13000 bytes
Requests per second:    2122.08 [#/sec] (mean)
Time per request:       47.124 [ms] (mean)
Time per request:       0.471 [ms] (mean, across all concurrent requests)
Transfer rate:          426.90 [Kbytes/sec] received
```

Mako 10 on FrankenPHP + PHP 8.3:

```
Concurrency Level:      100
Time taken for tests:   0.118 seconds
Complete requests:      1000
Failed requests:        0
Total transferred:      169000 bytes
HTML transferred:       13000 bytes
Requests per second:    8491.27 [#/sec] (mean)
Time per request:       11.777 [ms] (mean)
Time per request:       0.118 [ms] (mean, across all concurrent requests)
Transfer rate:          1401.39 [Kbytes/sec] received
```

**Reply by n0nag0n** [reply]
That's awesome! I love those kinds of results! Changing very little code but getting 4x the output!
